AI/BattleAI/AttackPossibility.cpp

```diff
@@ -1,97 +1,188 @@
-/*
- * AttackPossibility.cpp, part of VCMI engine
- *
- * Authors: listed in file AUTHORS in main folder
- *
- * License: GNU General Public License v2.0 or later
- * Full text of license available in license.txt file, in main folder
- *
- */
-#include "StdInc.h"
-#include "AttackPossibility.h"
-
-AttackPossibility::AttackPossibility(BattleHex tile_, const BattleAttackInfo & attack_)
-	: tile(tile_),
-	attack(attack_)
-{
-}
-
-
-int64_t AttackPossibility::damageDiff() const
-{
-	//TODO: use target priority from HypotheticBattle
-	const auto dealtDmgValue = damageDealt;
-	const auto receivedDmgValue = damageReceived;
-
-	int64_t diff = 0;
-
-	//friendly fire or not
-	if(attack.attacker->unitSide() == attack.defender->unitSide())
-		diff = -dealtDmgValue - receivedDmgValue;
-	else
-		diff = dealtDmgValue - receivedDmgValue;
-
-	//mind control
-	auto actualSide = getCbc()->playerToSide(getCbc()->battleGetOwner(attack.attacker));
-	if(actualSide && actualSide.get() != attack.attacker->unitSide())
-		diff = -diff;
-	return diff;
-}
-
-int64_t AttackPossibility::attackValue() const
-{
-	return damageDiff() + tacticImpact;
-}
-
-AttackPossibility AttackPossibility::evaluate(const BattleAttackInfo & attackInfo, BattleHex hex)
-{
-	const std::string cachingStringBlocksRetaliation = "type_BLOCKS_RETALIATION";
-	static const auto selectorBlocksRetaliation = Selector::type(Bonus::BLOCKS_RETALIATION);
-
-	const bool counterAttacksBlocked = attackInfo.attacker->hasBonus(selectorBlocksRetaliation, cachingStringBlocksRetaliation);
-
-	AttackPossibility ap(hex, attackInfo);
-
-	ap.attackerState = attackInfo.attacker->acquireState();
-
-	const int totalAttacks = ap.attackerState->getTotalAttacks(attackInfo.shooting);
-
-	if(!attackInfo.shooting)
-		ap.attackerState->setPosition(hex);
-
-	auto defenderState = attackInfo.defender->acquireState();
-	ap.affectedUnits.push_back(defenderState);
-
-	for(int i = 0; i < totalAttacks; i++)
+/*
+ * AttackPossibility.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "StdInc.h"
+#include "AttackPossibility.h"
+#include "../../lib/CStack.h"//todo: remove

```

Comment:
Could you explain in the comment why do you want to remove this include?

Reply:
This is my old comment, copied from `BattleAI.cpp` it is still relevant even here: AI should not use `CStack` directly, only `battle::Unit`, even `CUnitState` should be avoided.

Reply:
I suggest to put this information directly into the comment so that nobody needs to ask this question again. But still why can not we use it? Does it have some information normally hidden from the player or does it have some inactual data?

Reply:
It`s OK that people ask about it, it remind me to finish battle code rewrite :) Eventually only `IBattleInfoCallback` and `battle::Unit` should be used in code, CUnitState should be private and CStack should be removed completely.